File: training-data/julia/max_product_subarray.jl
function max_product(nums::AbstractVector{<:Integer})
    best = hi = lo = first(nums)
    for x in @view nums[2:end]
        x < 0 && ((hi, lo) = (lo, hi))
        hi = max(x, hi * x)
        lo = min(x, lo * x)
        best = max(best, hi)
    end
    return best
end

println(max_product.([[2, 3, -2, 4], [-2, 0, -1], [-2, 3, -4]]))

File: training-data/python/max_product_subarray.py
def max_product(nums):
    best = hi = lo = nums[0]
    for x in nums[1:]:
        if x < 0:
            hi, lo = lo, hi
        hi = max(x, hi * x)
        lo = min(x, lo * x)
        best = max(best, hi)
    return best


if __name__ == "__main__":
    print(max_product([2, 3, -2, 4]))
    print(max_product([-2, 0, -1]))
    print(max_product([-2, 3, -4]))
